### PR TITLE
Device const qualifier check and fix 

### PR DESCRIPTION
**Pull request by tbursztyka**
It's been a while since that has been checked.

### CODE REVIEW

**Comment by tbursztyka** [comment]
hum... looks like when I am adding a reviewer, it removes another one. Is there a limit in reviewers or my browser is just messing up with github's ui?

**Review by jfischer-no — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
> hum... looks like when I am adding a reviewer, it removes another one. Is there a limit in reviewers or my browser is just messing up with github's ui?

There's a limit of 15 reviewers. It's causing problems elsewhere too.

File: drivers/entropy/entropy_cc13xx_cc26xx.c [hunk 1]
```diff
@@ -131,7 +131,7 @@ static int entropy_cc13xx_cc26xx_get_entropy(const struct device *dev,
 	return 0;
 }
 
-static void entropy_cc13xx_cc26xx_isr(const void *arg)
+static void entropy_cc13xx_cc26xx_isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/entropy/entropy_cc13xx_cc26xx.c:134]:
```suggestion
static void entropy_cc13xx_cc26xx_isr(const struct device *dev)
```

File: drivers/entropy/entropy_nrf5.c [hunk 2]
```diff
@@ -204,7 +204,7 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size, uint8_t threshol
 	rngp->threshold	  = threshold;
 }
 
-static void isr(const void *arg)
+static void isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/entropy/entropy_nrf5.c:207]:
```suggestion
static void isr(const struct device *dev)
```

File: drivers/espi/host_subs_npcx.c [hunk 3]
```diff
@@ -223,7 +223,7 @@ static inline uint8_t host_shd_mem_wnd_size_sl(uint32_t size)
 
 /* Host KBC sub-device local functions */
 #if defined(CONFIG_ESPI_PERIPHERAL_8042_KBC)
-static void host_kbc_ibf_isr(void *arg)
+static void host_kbc_ibf_isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/espi/host_subs_npcx.c:226]:
```suggestion
static void host_kbc_ibf_isr(const struct device *dev)
```

File: drivers/espi/host_subs_npcx.c [hunk 4]
```diff
@@ -438,7 +438,7 @@ static void host_shared_mem_region_init(void)
 #if defined(CONFIG_ESPI_PERIPHERAL_HOST_IO) || \
 				defined(CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD)
 /* Host pm (host io) sub-module isr function for all channels such as ACPI. */
-static void host_pmch_ibf_isr(void *arg)
+static void host_pmch_ibf_isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/espi/host_subs_npcx.c:441]:
```suggestion
static void host_pmch_ibf_isr(const struct device *dev)
```

File: drivers/espi/host_subs_npcx.c [hunk 5]
```diff
@@ -438,7 +438,7 @@ static void host_shared_mem_region_init(void)
...
 {
 	ARG_UNUSED(arg);
```
Comment (henrikbrixandersen) [inline drivers/espi/host_subs_npcx.c:443]:
```suggestion
	ARG_UNUSED(dev);
```

File: drivers/espi/host_subs_npcx.c [hunk 6]
```diff
@@ -249,7 +249,7 @@ static void host_kbc_ibf_isr(void *arg)
 							evt);
 }
 
-static void host_kbc_obe_isr(void *arg)
+static void host_kbc_obe_isr(const struct device *arg)
 {
 	ARG_UNUSED(arg);
```
Comment (henrikbrixandersen) [inline drivers/espi/host_subs_npcx.c:254]:
```suggestion
	ARG_UNUSED(dev);
```

File: drivers/gpio/gpio_lpc11u6x.c [hunk 7]
```diff
@@ -418,7 +418,7 @@ static uint32_t gpio_lpc11u6x_get_pending_int(const struct device *dev)
 	return -ENOTSUP;
 }
 
-static void gpio_lpc11u6x_isr(const void *arg)
+static void gpio_lpc11u6x_isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_lpc11u6x.c:421]:
```suggestion
static void gpio_lpc11u6x_isr(const struct device *dev)
```

File: drivers/gpio/gpio_lpc11u6x.c [hunk 8]
```diff
@@ -418,7 +418,7 @@ static uint32_t gpio_lpc11u6x_get_pending_int(const struct device *dev)
...
 {
 	struct gpio_lpc11u6x_shared *shared =
 		(struct gpio_lpc11u6x_shared *)arg;
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_lpc11u6x.c:424]: This looks weird.

File: drivers/i2c/i2c_cc13xx_cc26xx.c [hunk 9]
```diff
@@ -278,7 +278,7 @@ static int i2c_cc13xx_cc26xx_configure(const struct device *dev,
 	return 0;
 }
 
-static void i2c_cc13xx_cc26xx_isr(const void *arg)
+static void i2c_cc13xx_cc26xx_isr(const struct device *arg)
```
Comment (henrikbrixandersen) [inline drivers/i2c/i2c_cc13xx_cc26xx.c:281]:
```suggestion
static void i2c_cc13xx_cc26xx_isr(const struct device *dev)
```

File: drivers/i2c/i2c_cc13xx_cc26xx.c [hunk 10]
```diff
@@ -278,7 +278,7 @@ static int i2c_cc13xx_cc26xx_configure(const struct device *dev,
...
 {
 	const uint32_t base = get_dev_config(arg)->base;
 	struct i2c_cc13xx_cc26xx_data *data = get_dev_data(arg);
```
Comment (henrikbrixandersen) [inline drivers/i2c/i2c_cc13xx_cc26xx.c:284]:
```suggestion
	const uint32_t base = get_dev_config(dev)->base;
	struct i2c_cc13xx_cc26xx_data *data = get_dev_data(dev);
```

File: drivers/i2c/i2c_mchp_xec.c [hunk 11]
```diff
@@ -457,7 +457,7 @@ static int i2c_xec_transfer(const struct device *dev, struct i2c_msg *msgs,
 	return 0;
 }
 
-static void i2c_xec_bus_isr(void *arg)
+static void i2c_xec_bus_isr(const struct device *arg)
 {
 #ifdef CONFIG_I2C_SLAVE
 	struct device *dev = (struct device *)arg;
```
Comment (henrikbrixandersen) [inline drivers/i2c/i2c_mchp_xec.c:463]: This should be cleaned up, imho.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
I started commenting all the places where the `arg` name is kept for the ISR function argument to replace it with `dev`, but gave up seeing this is all over the place.

I would like to see all those `const struct device *arg` be renamed to `const struct device *dev` as that will trigger clean-up of the remaining code in the affected ISR functions (where we now have a `const struct device *arg` being typecasted to `const struct device *dev` and other, similar constructs.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 12]
```diff
@@ -868,7 +868,7 @@ static int i2c_it8xxx2_transfer(const struct device *dev, struct i2c_msg *msgs,
 	return i2c_parsing_return_value(dev);
 }
 
-static void i2c_it8xxx2_isr(void *arg)
+static void i2c_it8xxx2_isr(const struct device *arg)
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/i2c/i2c_ite_it8xxx2.c:871]: That kind of change my script does not do :)
I could probably add another rule to change the param name

File: drivers/i2c/i2c_mchp_xec.c (hunk 11 above)
Comment (tbursztyka) [reply drivers/i2c/i2c_mchp_xec.c:463]: the script  could not match this case. I'll update it.

Comment (tbursztyka) [reply drivers/i2c/i2c_mchp_xec.c:463]: looks like cocci does not like the #ifdef. it should be part of the '...' statement, but no.

**Review by brockus-zephyr — DISMISSED** [review]

File: tests/subsys/pm/power_mgmt/src/main.c [hunk 13]
```diff
@@ -411,15 +411,15 @@ void test_busy(void)
...
+	pm_device_state_lock((const struct device *)device_a);
 	zassert_true(pm_device_state_is_locked(device_a), NULL);
 
 	testing_device_lock = true;
 	enter_low_power = true;
 
 	k_sleep(SLEEP_TIMEOUT);
 
-	pm_device_state_unlock((struct device *)device_a);
+	pm_device_state_unlock((const struct device *)device_a);
```
Comment (gmarull) [inline tests/subsys/pm/power_mgmt/src/main.c:422]: `device_a` is already a `const struct device*`

**Review by erwango — DISMISSED** [review]

**Review by ruuddw — DISMISSED** [review]
ok for the ARC change, thanks.

File: drivers/counter/counter_esp32.c [hunk 14]
```diff
@@ -238,7 +238,7 @@ static const struct counter_driver_api counter_api = {
 
 static void counter_esp32_isr(void *arg)
```
Comment (henrikbrixandersen) [inline drivers/counter/counter_esp32.c:239]:
```suggestion
static void counter_esp32_isr(const struct device *dev)
```

File: drivers/i2c/i2c_esp32.c [hunk 15]
```diff
@@ -593,7 +593,7 @@ static int IRAM_ATTR i2c_esp32_transfer(const struct device *dev, struct i2c_msg
 
 static void IRAM_ATTR i2c_esp32_isr(void *arg)
```
Comment (henrikbrixandersen) [inline drivers/i2c/i2c_esp32.c:594]:
```suggestion
static void IRAM_ATTR i2c_esp32_isr(const struct device *dev)
```

File: drivers/interrupt_controller/intc_gd32_exti.c [hunk 16]
```diff
@@ -82,7 +82,7 @@ static const uint8_t line2irq[NUM_EXTI_LINES] = {
 #endif /* CONFIG_SOC_SERIES_GD32F4XX */
 };
 
-__unused static void gd32_exti_isr(void *isr_data)
+__unused static void gd32_exti_isr(const void *isr_data)
```
Comment (henrikbrixandersen) [inline drivers/interrupt_controller/intc_gd32_exti.c:85]: Was this intended?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_esp32.c (hunk 14 above)
Comment (tbursztyka) [reply drivers/counter/counter_esp32.c:239]: This one could not be found by the script, as it does note use IRQ_CONNECT macro.

File: drivers/interrupt_controller/intc_gd32_exti.c (hunk 16 above)
Comment (tbursztyka) [reply drivers/interrupt_controller/intc_gd32_exti.c:85]: the script works by steps, and it first checks about const qualifier on isr used through IRQ_CONNECT, so yes it's fine.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by tbursztyka** [comment]
Commit d2118eb1bcb has flaws, I wonder how this passed CI as it is blocking mine...

**Comment by gmarull** [comment]
> Commit [d2118eb](https://github.com/zephyrproject-rtos/zephyr/commit/d2118eb1bcb3bf991c076cad95fc5f0176785c9e) has flaws, I wonder how this passed CI as it is blocking mine...

https://github.com/zephyrproject-rtos/zephyr/pull/41626

**Comment by erwango** [comment]
> > Commit [d2118eb](https://github.com/zephyrproject-rtos/zephyr/commit/d2118eb1bcb3bf991c076cad95fc5f0176785c9e) has flaws, I wonder how this passed CI as it is blocking mine...
> 
> #41626

I guess you meant https://github.com/zephyrproject-rtos/zephyr/pull/41999

File: drivers/i2c/i2c_esp32.c [hunk 17]
```diff
@@ -591,9 +591,8 @@ static int IRAM_ATTR i2c_esp32_transfer(const struct device *dev, struct i2c_msg
 	return ret;
 }
 
-static void IRAM_ATTR i2c_esp32_isr(void *arg)
+static void IRAM_ATTR i2c_esp32_isr(const struct device *dev)
```
Comment (sylvioalves) [inline drivers/i2c/i2c_esp32.c:594]: Both changes in ESP32 (counter and I2C) will add build warnings with this change. `esp_intr_alloc` requires `void *arg` in its signature. Can this be removed from this PR?

Comment (tbursztyka) [reply drivers/i2c/i2c_esp32.c:594]: sure

**Review by sylvioalves — DISMISSED** [review]

File: drivers/can/can_stm32fd.c [hunk 18]
```diff
@@ -152,9 +152,8 @@ static int can_stm32fd_set_timing(const struct device *dev,
 	return can_mcan_set_timing(mcan_cfg, timing, timing_data);
 }
 
-static void can_stm32fd_line_0_isr(void *arg)
+void can_stm32fd_line_0_isr(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/can/can_stm32fd.c:155]: Did you mean to drop `static`?

File: drivers/can/can_stm32fd.c [hunk 19]
```diff
@@ -164,11 +163,9 @@ static void can_stm32fd_line_0_isr(void *arg)
 	can_mcan_line_0_isr(mcan_cfg, msg_ram, mcan_data);
 }
 
-static void can_stm32fd_line_1_isr(void *arg)
+void can_stm32fd_line_1_isr(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/can/can_stm32fd.c:166]: Did you mean to drop `static`?

File: drivers/can/can_stm32fd.c (hunk 18 above)
Comment (tbursztyka) [reply drivers/can/can_stm32fd.c:155]: good catch, no indeed. Will check, that's not supposed to happen obviously

File: arch/arc/core/arc_smp.c [hunk 20]
```diff
@@ -114,7 +114,7 @@ void z_arc_slave_start(int cpu_num)
 
 #ifdef CONFIG_SMP
 
-static void sched_ipi_handler(const void *unused)
+static void sched_ipi_handler(const struct device *unused)
```
Comment (nashif) [inline arch/arc/core/arc_smp.c:117]: not following how this ends up a struct device, it is unrelated to devices...

**Review by nashif — CHANGES_REQUESTED** [review]

File: arch/arm64/core/smp.c [hunk 21]
```diff
@@ -166,7 +166,7 @@ static void broadcast_ipi(unsigned int ipi)
 	gic_raise_sgi(ipi, mpidr, SGIR_TGT_MASK & ~(1 << MPIDR_TO_CORE(mpidr)));
 }
 
-void sched_ipi_handler(const void *unused)
+void sched_ipi_handler(const struct device *unused)
```
Comment (carlocaione) [inline arch/arm64/core/smp.c:169]: same here & co.

File: arch/arc/core/arc_smp.c (hunk 20 above)
Comment (tbursztyka) [reply arch/arc/core/arc_smp.c:117]: Looks like I need to improve the cocci rule: if the parameter does not get casted to struct device into the function, it should not match.

File: drivers/timer/arm_arch_timer.c [hunk 22]
```diff
@@ -18,9 +18,9 @@
 static struct k_spinlock lock;
 static uint64_t last_cycle;
 
-static void arm_arch_timer_compare_isr(const void *arg)
+static void arm_arch_timer_compare_isr(const struct device *dev)
```
Comment (carlocaione) [inline drivers/timer/arm_arch_timer.c:21]: This is not a device.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/counter/counter_native_posix.c [hunk 23]
```diff
@@ -22,9 +22,9 @@ static struct counter_alarm_cfg pending_alarm;
 static bool is_alarm_pending;
 static const struct device *device;
 
-static void counter_isr(const void *arg)
+static void counter_isr(const struct device *dev)
```
Comment (carlocaione) [inline drivers/counter/counter_native_posix.c:25]: again. Why?

File: drivers/entropy/entropy_nrf5.c [hunk 24]
```diff
@@ -201,11 +201,11 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size, uint8_t threshol
 	rngp->threshold	  = threshold;
 }
 
-static void isr(const void *arg)
+static void isr(const struct device *dev)
```
Comment (carlocaione) [inline drivers/entropy/entropy_nrf5.c:204]: same

File: drivers/entropy/entropy_stm32.c [hunk 25]
```diff
@@ -278,11 +278,11 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size,
 	rngp->threshold	  = threshold;
 }
 
-static void stm32_rng_isr(const void *arg)
+static void stm32_rng_isr(const struct device *dev)
```
Comment (carlocaione) [inline drivers/entropy/entropy_stm32.c:281]: same

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: tests/kernel/gen_isr_table/src/main.c [hunk 26]
```diff
@@ -100,14 +100,14 @@ ISR_DIRECT_DECLARE(isr2)
 }
 #endif
 
-void isr3(const void *param)
+void isr3(const struct device *param)
```
Comment (carlocaione) [inline tests/kernel/gen_isr_table/src/main.c:103]: For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

Comment (carlocaione) [inline tests/kernel/gen_isr_table/src/main.c:103]: For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

Comment (carlocaione) [inline tests/kernel/gen_isr_table/src/main.c:103]: For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

Comment (mnkp) [reply tests/kernel/gen_isr_table/src/main.c:103]: The parameter passed to the ISR is defined in `IRQ_CONNECT` macro [Defining a regular ISR](https://docs.zephyrproject.org/latest/reference/kernel/other/interrupts.html#defining-a-regular-isr). Most of the time drivers pass `struct device` but in a few cases that you pointed out that doesn't seem to be a case so it indeed should be fixed.

Comment (tbursztyka) [reply tests/kernel/gen_isr_table/src/main.c:103]: and the script I run did not check that the actual parameter is being used an actual struct device then. Now it's fixed.

**Review by carlocaione — DISMISSED** [review]
The obvious problems were addressed. I still do not really like the fact that we are specializing the ISR argument to the actual usage, but I think that is a matter of personal preference.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by tbursztyka** [comment]
@nashif  please, re-review

**Comment by henrikbrixandersen** [comment]
@tbursztyka Please rebase and resolve the merge conflicts.

**Review by nashif — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]
PM bits LGTM

**Comment by mbolivar-nordic** [comment]
This needs another rebase already :(

**Comment by nashif** [comment]
> This needs another rebase already :( 

it has been a while since the last push :)

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]

File: drivers/i2c/i2c_dw.c [hunk 27]
```diff
@@ -197,9 +197,8 @@ static inline void i2c_dw_write_byte_non_blocking(const struct device *dev, uint
 static void i2c_dw_slave_read_clear_intr_bits(const struct device *dev);
 #endif
 
-static void i2c_dw_isr(void *arg)
+static void i2c_dw_isr(const struct device *port)
```
Comment (nashif) [inline drivers/i2c/i2c_dw.c:200]: dev?

**Review by jfischer-no — DISMISSED** [review]

File: drivers/i2c/i2c_dw.c (hunk 27 above)
Comment (tbursztyka) [reply drivers/i2c/i2c_dw.c:200]: "port" was kept as this is the name used already for the device for the code (see below: arg was casted to a struct device *port).

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**